terraform-week-4/main.tf

```diff
@@ -0,0 +1,252 @@
+resource "aws_security_group" "ssh" {
+  name = "ssh"
+  description = "Allow incoming SSH connections."
+  vpc_id = aws_vpc.prod.id
+
+  ingress {
+    from_port = 22
+    to_port = 22
+    protocol = "tcp"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+
+  egress {
+    from_port = 0
+    to_port = 0
+    protocol = "-1"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+}
+
+resource "aws_security_group" "http" {
+  name = "http"
+  description = "Allow incoming HTTP connections."
+  vpc_id = aws_vpc.prod.id
+
+  ingress {
+    from_port = 80
+    to_port = 80
+    protocol = "tcp"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+  ingress {
+    from_port = 443
+    to_port = 443
+    protocol = "tcp"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+  egress {
+    from_port = 0
+    to_port = 0
+    protocol = "-1"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+}
+
+resource "aws_subnet" "public" {
+  vpc_id = aws_vpc.prod.id
+  cidr_block = "10.0.1.0/24"
+  availability_zone = "us-west-2b"
+  map_public_ip_on_launch = true
+
+  tags = {
+    Name = "Public"
+  }
+}
+
+resource "aws_subnet" "private" {
+  vpc_id = aws_vpc.prod.id
+  cidr_block = "10.0.2.0/24"
+  availability_zone = "us-west-2a"
+
+  tags = {
+    Name = "Private"
+  }
+}
+resource "aws_instance" "web_public" {
+  ami = "${lookup(var.amis, var.aws_region)}"
+  instance_type = "t2.micro"
+  key_name = "${var.aws_key_name}"
+  subnet_id = aws_subnet.public.id
+
+  security_groups = [aws_security_group.ssh.id, aws_security_group.http.id]
+  user_data = file("script.sh")
+  depends_on = [aws_security_group.http, aws_security_group.ssh]
+}
+
+resource "aws_instance" "web_private" {
+  ami = "${lookup(var.amis, var.aws_region)}"
+  instance_type = "t2.micro"
+  key_name = "${var.aws_key_name}"
+  subnet_id = aws_subnet.private.id
+  security_groups = [aws_security_group.ssh.id, aws_security_group.http.id]
+  user_data = file("script-private.sh")
+  depends_on = [aws_security_group.http, aws_security_group.ssh]
+}
+
+resource "aws_vpc" "prod" {
+  cidr_block = "10.0.0.0/16"
+  tags = {
+    Name = "prod vpc"
+  }
+}
+
+resource "aws_internet_gateway" "gw_main" {
+  vpc_id = aws_vpc.prod.id
+}
+
+resource "aws_route_table" "prod" {
+  vpc_id = aws_vpc.prod.id
+
+  route {
+    cidr_block = "0.0.0.0/0"
+    gateway_id = aws_internet_gateway.gw_main.id
+  }
+
+  route {
+    ipv6_cidr_block = "::/0"
+    gateway_id = aws_internet_gateway.gw_main.id
+  }
+
+  tags = {
+    Name = "prod"
+  }
+}
+
+resource "aws_route_table_association" "prod" {
+  subnet_id = aws_subnet.public.id
+  route_table_id = aws_route_table.prod.id
+}
+
+resource "aws_eip" "prod" {
+  vpc = true
+}
+
+resource "aws_nat_gateway" "gw_nat" {
```

Comment:
The `aws_nat_gateway` service is not included to Free Tier, so it always costs.
To avoid that we recommend to use Nat instance instead which is actually just EC2 instance and it is included to Free Tier